File: frameworks/bean.core/src/main/java/com/example/scaffold/frameworks/config/center/PayloadHandlerMethodArgumentResolver.java
package com.example.scaffold.frameworks.config.center;

import com.example.scaffold.frameworks.valueobjects.payload.Payload;
import org.springframework.core.MethodParameter;
import org.springframework.http.converter.HttpMessageConverter;
import org.springframework.lang.NonNull;
import org.springframework.web.bind.support.WebDataBinderFactory;
import org.springframework.web.context.request.NativeWebRequest;
import org.springframework.web.method.support.ModelAndViewContainer;
import org.springframework.web.servlet.mvc.method.annotation.AbstractMessageConverterMethodProcessor;

import java.util.LinkedHashMap;
import java.util.List;

public class PayloadHandlerMethodArgumentResolver extends AbstractMessageConverterMethodProcessor {

    protected PayloadHandlerMethodArgumentResolver(List<HttpMessageConverter<?>> converters) {
        super(converters);
    }

    @Override
    public boolean supportsParameter(MethodParameter parameter) {
        return parameter.getParameterType().isAssignableFrom(Payload.class);
    }

    @Override
    public Object resolveArgument(@NonNull MethodParameter parameter, ModelAndViewContainer mavContainer, @NonNull NativeWebRequest webRequest,
                                  WebDataBinderFactory binderFactory) throws Exception {
        Object arg = readWithMessageConverters(webRequest, parameter, LinkedHashMap.class.getGenericSuperclass());
        return new Payload(arg);
    }

    @Override
    public boolean supportsReturnType(@NonNull MethodParameter returnType) {
        return false;
    }

    @Override
    public void handleReturnValue(Object returnValue, @NonNull MethodParameter returnType, @NonNull ModelAndViewContainer mavContainer,
                                  @NonNull NativeWebRequest webRequest) {
    }
}

File: frameworks/bean.core/src/main/kotlin/com/qomolangma/frameworks/bean/core/PayloadHandlerMethodArgumentResolver.kt
package com.qomolangma.frameworks.bean.core

import com.qomolangma.frameworks.payload.core.Payload
import lombok.Generated
import org.springframework.core.MethodParameter
import org.springframework.http.converter.HttpMessageConverter
import org.springframework.web.bind.support.WebDataBinderFactory
import org.springframework.web.context.request.NativeWebRequest
import org.springframework.web.method.support.ModelAndViewContainer
import org.springframework.web.servlet.mvc.method.annotation.AbstractMessageConverterMethodProcessor

open class PayloadHandlerMethodArgumentResolver constructor(converters: List<HttpMessageConverter<*>>) :
    AbstractMessageConverterMethodProcessor(converters) {
    override fun supportsParameter(parameter: MethodParameter): Boolean {
        return parameter.parameterType.isAssignableFrom(Payload::class.java)
    }

    @Throws(Exception::class)
    @Generated
    override fun resolveArgument(
        parameter: MethodParameter,
        mavContainer: ModelAndViewContainer?,
        webRequest: NativeWebRequest,
        binderFactory: WebDataBinderFactory?
    ): Any? {
        return readWithMessageConverters<Any>(
            webRequest,
            parameter,
            LinkedHashMap::class.java.genericSuperclass
        )?.let {
            Payload(
                it
            )
        }
    }

    override fun supportsReturnType(returnType: MethodParameter): Boolean = false

    override fun handleReturnValue(
        returnValue: Any?,
        returnType: MethodParameter,
        mavContainer: ModelAndViewContainer,
        webRequest: NativeWebRequest
    ) {
    }
}
